src/lib/scheduler/operator_task.cpp

```diff
@@ -10,48 +10,62 @@
 #include "scheduler/job_task.hpp"
 #include "utils/tracing/probes.hpp"
 
-namespace opossum {
-OperatorTask::OperatorTask(std::shared_ptr<AbstractOperator> op, SchedulePriority priority, bool stealable)
-    : AbstractTask(priority, stealable), _op(std::move(op)) {}
-
-std::string OperatorTask::description() const {
-  return "OperatorTask with id: " + std::to_string(id()) + " for op: " + _op->description();
-}
-
-std::vector<std::shared_ptr<AbstractTask>> OperatorTask::make_tasks_from_operator(
-    const std::shared_ptr<AbstractOperator>& op) {
-  std::vector<std::shared_ptr<AbstractTask>> tasks;
-  std::unordered_map<std::shared_ptr<AbstractOperator>, std::shared_ptr<AbstractTask>> task_by_op;
-  _add_tasks_from_operator(op, tasks, task_by_op);
-  return tasks;
-}
-
-std::shared_ptr<AbstractTask> OperatorTask::_add_tasks_from_operator(
-    const std::shared_ptr<AbstractOperator>& op, std::vector<std::shared_ptr<AbstractTask>>& tasks,
-    std::unordered_map<std::shared_ptr<AbstractOperator>, std::shared_ptr<AbstractTask>>& task_by_op) {
-  const auto task_by_op_it = task_by_op.find(op);
-  if (task_by_op_it != task_by_op.end()) return task_by_op_it->second;
-
-  auto task = std::make_shared<OperatorTask>(op);
-  task_by_op.emplace(op, task);
+namespace {
+
+using namespace opossum;  // NOLINT
+
+/**
+ * Create tasks recursively. Called by `make_tasks_from_operator`.
+ * @returns the root of the subtree that was added.
+ * @param task_by_op  Cache to avoid creating duplicate Tasks for diamond shapes
+ */
+std::shared_ptr<AbstractTask> add_tasks_from_operator_recursively(const std::shared_ptr<AbstractOperator>& op,
+                                                                  std::vector<std::shared_ptr<AbstractTask>>& tasks) {
+  auto task = op->operator_task();
+  if (!task) {
```

Comment:
This is not thread-safe. In 95% of the cases, a conditional after an atomic read smells like trouble. Maybe `op->get_or_create_operator_task`?

Reply:
Clever, thanks! I missed that one.